src/rgw/rgw_rados.cc

```diff
@@ -6306,77 +6306,451 @@ int RGWRados::Object::Read::read(int64_t ofs, int64_t end, bufferlist& bl, optio
+  struct aiocb *cb = new struct aiocb;
+  memset(cb, 0, sizeof(struct aiocb));
+  cb->aio_fildes = ::open(location.c_str(), O_RDONLY);
+  if (cb->aio_fildes < 0)  {
+    ldout(cct, 0) << "Error: add_cache_io:: read ::open the file has return error "  << dendl;
+    return -1;
+  }
+
+  cb->aio_buf = malloc(len);
+  cb->aio_nbytes = len;
+  cb->aio_offset = read_ofs;
+  cb->aio_sigevent.sigev_notify = SIGEV_THREAD;
+  cb->aio_sigevent.sigev_notify_function = _cache_aio_completion_cb;
+  cb->aio_sigevent.sigev_notify_attributes = NULL;
+  cb->aio_sigevent.sigev_value.sival_ptr = (void*)c;
+  c->paiocb = cb;
+
+  cache_lock.lock();
+  cache_aio_map[ofs] = c;
+  cache_lock.unlock();
+  *cc = c;
+  return 0;
+}
+
+int get_obj_data::submit_l1_io_read(bufferlist *bl, int len, string oid) {
+
+  std::string location = cct->_conf->rgw_datacache_persistent_path + oid; 
+  int cache_file = -1;
+  int r = 0;
+
+  cache_file = ::open(location.c_str(), O_RDONLY);
+  if (cache_file < 0) {
+    dout(0) << "ERROR: get_obj_data::cache_io_read, open failed. errno " << errno << dendl;
+    return -1;
   }
 
-  int drain() {
-    auto c = aio->wait();
-    while (!c.empty()) {
-      int r = flush(std::move(c));
-      if (r < 0) {
-        cancel();
-        return r;
-      }
-      c = aio->wait();
-    }
-    return flush(std::move(c));
+  r = ::read(cache_file, tmp_data, len);
+  if (r < 0){
+    dout(0) << "ERROR: get_obj_data::cache_io_read, read failed. errno " << r << dendl;
+    goto END;
   }
-};
+  bl->append(tmp_data, len);
+
+END:
+  ::close(cache_file);
+  return r;
+}
+
+int get_obj_data::submit_l1_aio_read(L1CacheRequest *cc) {
+
+  int r = 0;
+  if((r= ::aio_read(cc->paiocb)) != 0) {
+    ldout(cct, 0) << "ERROR: aio_read ::aio_read"<< r << dendl;
+  }
+  return r;
+}
+
+void _cache_aio_completion_cb(sigval_t sigval)
+{ 
+  CacheRequest *c = static_cast<CacheRequest *>(sigval.sival_ptr);
+  c->op_data->cache_aio_completion_cb(c);
+}
+
+void get_obj_data::cache_aio_completion_cb(CacheRequest *c){
+
+  int status = c->status();
+  if (status == ECANCELED) {
+    cache_unmap_io(c->ofs);
+    ldout(cct, 0) << "DataCache: cache_aio_request: status = ECANCLE" << dendl;
+    return;
+  } else if (status == 0) {
+    cache_unmap_io(c->ofs);
+    l2_lock.lock();
+    ldout(cct, 0) << "DataCache: cache_aio_completion_cb,inside the finish lock- oid : " << c->oid << ", len:" << std::hex << c->len << "plb->len" << std::hex << c->pbl->length()  << dendl;
+    c->finish();
+    l2_lock.unlock();
+  }
+}
+
+void get_obj_data::cache_unmap_io(off_t ofs){
+	
+  cache_lock.lock();
+  map<off_t, struct CacheRequest *>::iterator iter = cache_aio_map.find(ofs);
+  if (iter == cache_aio_map.end()) {
+    cache_lock.unlock();
+    return;
+  }
+  cache_aio_map.erase(ofs);
+  cache_lock.unlock();
+}
+
+int get_obj_data::add_cache_notifier(std::string oid, librados::AioCompletion *lc) {
+  //Context *onack = new IoCtxImpl::C_aio_Complete(lc->pc);
```

Comment:
what's going on with the commented code here--this function is now a no-op?

Reply:
> what's going on with the commented code here--this function is now a no-op?

yes it's a no-op, removed this function